File: engine/compose/src/commonMain/kotlin/dev/getmilano/core/MilanoIdentifier.kt
package dev.getmilano

/**
 * Unicode scalar count, in pure Kotlin: surrogate pairs count once. The
 * document model's lengths (expression limit, `length()`) are defined in
 * scalars, never UTF-16 units or grapheme clusters.
 */
internal fun String.unicodeScalarCount(): Int {
    var count = 0
    var index = 0
    while (index < length) {
        val c = this[index]
        index +=
            if (c.isHighSurrogate() && index + 1 < length &&
                this[index + 1].isLowSurrogate()
            ) {
                2
            } else {
                1
            }
        count++
    }
    return count
}

/**
 * The one identifier grammar for component types, properties, events,
 * actions, and state and context keys: a letter followed by letters,
 * digits, or underscores. Case-sensitive; never starts with `$`.
 */
internal object MilanoIdentifier {
    fun isValid(name: String): Boolean {
        if (name.isEmpty()) return false
        val first = name[0]
        if (first !in 'a'..'z' && first !in 'A'..'Z') return false
        return name.drop(1).all { it in 'a'..'z' || it in 'A'..'Z' || it in '0'..'9' || it == '_' }
    }
}

File: engine/swiftui/Sources/MilanoSDK/Core/MilanoIdentifier.swift
import Foundation

/// The one identifier grammar for component types, properties, events,
/// actions, and state and context keys: a letter followed by letters,
/// digits, or underscores. Case-sensitive; never starts with `$`.
enum MilanoIdentifier {
    static func isValid(_ name: String) -> Bool {
        guard let first = name.unicodeScalars.first else { return false }
        guard (first >= "a" && first <= "z") || (first >= "A" && first <= "Z") else { return false }
        return name.unicodeScalars.dropFirst().allSatisfy { scalar in
            (scalar >= "a" && scalar <= "z") || (scalar >= "A" && scalar <= "Z")
                || (scalar >= "0" && scalar <= "9") || scalar == "_"
        }
    }
}
